File: src/main/java/com/smartwallet/entity/Transaction.java
package com.smartwallet.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity(name = "PortfolioTransaction")
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_id", nullable = false)
    @JsonIgnore
    private Asset asset;

    @Enumerated(EnumType.STRING)
    @Column(name = "transaction_type", nullable = false, length = 20)
    private TransactionType transactionType;

    @Column(precision = 18, scale = 8, nullable = false)
    private BigDecimal quantity;

    @Column(precision = 15, scale = 2, nullable = false)
    private BigDecimal price;

    @Column(name = "total_value", precision = 15, scale = 2, nullable = false)
    private BigDecimal totalValue;

    @Column(precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal fees = BigDecimal.ZERO;

    @Column(name = "total_with_fees", precision = 15, scale = 2)
    private BigDecimal totalWithFees;

    @Column(name = "transaction_date", nullable = false)
    private LocalDateTime transactionDate;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (transactionDate == null) {
            transactionDate = LocalDateTime.now();
        }
        if (totalWithFees == null) {
            totalWithFees = totalValue.add(fees);
        }
    }

    public enum TransactionType {
        BUY,   // Compra
        SELL,  // Venda
        DIVIDEND, // Dividendos/Juros
        SPLIT, // Desdobramento
        MERGE  // Incorporação
    }
}

File: src/main/kotlin/com/smartwallet/portfolio/model/Transaction.kt
package com.smartwallet.portfolio.model

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.Instant

@Entity
@Table(name = "transactions")
data class Transaction(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_id", nullable = false)
    val asset: PortfolioAsset,

    @Column(name = "transaction_type", nullable = false, length = 20)
    val transactionType: String,

    @Column(nullable = false)
    val quantity: BigDecimal,

    @Column(nullable = false)
    val price: BigDecimal,

    @Column(name = "total_value", nullable = false)
    val totalValue: BigDecimal,

    val fees: BigDecimal? = BigDecimal.ZERO,

    @Column(name = "transaction_date", nullable = false)
    val transactionDate: Instant = Instant.now(),

    val notes: String? = null,

    @Column(name = "created_at")
    val createdAt: Instant? = null
